src/qml/editorwidgets/RelationEditor.qml

```diff
@@ -0,0 +1,239 @@
+
+        Item {
+          id: listitem
+          anchors { left: parent.left; right: parent.right }
+
+          focus: true
+
+          height: Math.max( itemHeight, featureText.height )
+
+          Text {
+            id: featureText
+            anchors { leftMargin: 10; left: parent.left; right: deleteButton.left; verticalCenter: parent.verticalCenter }
+            font.bold: true
+            color: readOnly ? "grey" : "black"
+            text: { text: model.displayString }
+          }
+
+          MouseArea {
+            anchors.fill: parent
+
+            onClicked: {
+                embeddedFeatureForm.state = !readOnly ? "Edit" : "ReadOnly"
+                embeddedFeatureForm.relatedFeature = associatedRelationId === "" ? model.referencingFeature : model.associatedReferencedFeature
+                embeddedFeatureForm.relatedLayer = associatedRelationId === "" ? relationEditorModel.relation.referencingLayer : relationEditorModel.associatedRelation.referencedLayer
+                embeddedFeatureForm.active = true
+            }
+          }
+
+          Row
+          {
+            id: editRow
+            anchors { top: parent.top; right: parent.right }
+            height: listitem.height
+
+            ToolButton {
+                id: deleteButton
+                width: parent.height
+                height: parent.height
+                visible: !readOnly
+
+                contentItem: Rectangle {
+                    anchors.fill: parent
+                    color: "black"
+                    Image {
+                      anchors.fill: parent
+                      anchors.margins: 4 * px
+                      fillMode: Image.PreserveAspectFit
+                      horizontalAlignment: Image.AlignHCenter
+                      verticalAlignment: Image.AlignVCenter
+                      source: Style.getThemeIcon( 'ic_delete_forever_white_24dp' )
+                    }
+                }
+
+                onClicked: {
+                    deleteDialog.referencingFeatureId = model.referencingFeature.id
+                    deleteDialog.visible = true
+                }
+            }
+          }
+
+          //bottom line
+          Rectangle {
+            id: bottomLine
+            anchors.bottom: parent.bottom
+            height: 1
+            color: "lightGray"
+            width: parent.width
+          }
+        }
+    }
+
+    //the delete entry stuff
+    MessageDialog {
+      id: deleteDialog
+
+      property int referencingFeatureId
+      property var layerName
+
+      visible: false
+
+      title: "Delete feature "+referencingFeatureId+" on layer "+layerName //translation needed
+      text: "Should the feature "+referencingFeatureId+" on layer "+layerName+" really be deleted?" //translation needed
+      standardButtons: StandardButton.Ok | StandardButton.Cancel
+      onAccepted: {
+        referencingFeatureListView.model.deleteFeature( referencingFeatureId )
```

Comment:
I wonder if we could move this to some central point in a utilities function or directly on QgsVectorLayer

Reply:
A VectorLayerUtils class that is registered as Qml Singleton maybe?